cura/PrinterOutputDevice.py

```diff
@@ -0,0 +1,303 @@
+from UM.OutputDevice.OutputDevice import OutputDevice
+from PyQt5.QtCore import pyqtProperty, pyqtSignal, pyqtSlot, QObject
+from enum import IntEnum  # For the connection state tracking.
+from UM.Logger import Logger
+
+
+##  Printer output device adds extra interface options on top of output device.
+#
+#   The assumption is made the printer is a FDM printer.
+#
+#   Note that a number of settings are marked as "final". This is because decorators
+#   are not inherited by children. To fix this we use the private counter part of those
+#   functions to actually have the implementation.
+#
+#   For all other uses it should be used in the same way as a "regular" OutputDevice.
+class PrinterOutputDevice(OutputDevice, QObject):
+    def __init__(self, device_id, parent = None):
+        QObject.__init__(self, parent)
+        OutputDevice.__init__(self, device_id)
+
+        self._target_bed_temperature = 0
+        self._bed_temperature = 0
+        self._num_extruders = 1
+        self._hotend_temperatures = [0] * self._num_extruders
+        self._target_hotend_temperatures = [0] * self._num_extruders
+        self._progress = 0
+        self._head_x = 0
+        self._head_y = 0
+        self._head_z = 0
+        self._connection_state = ConnectionState.CLOSED
+
+    def requestWrite(self, node, file_name = None, filter_by_machine = False):
+        raise NotImplementedError("requestWrite needs to be implemented")
+
+    # Signals:
+    bedTemperatureChanged = pyqtSignal()
+    targetBedTemperatureChanged = pyqtSignal()
+
+    progressChanged = pyqtSignal()
+
+    hotendTemperaturesChanged = pyqtSignal()
+    targetHotendTemperaturesChanged = pyqtSignal()
+
+    headPositionChanged = pyqtSignal()
+
+    connectionStateChanged = pyqtSignal(str)
+
+    ##  Get the bed temperature of the bed (if any)
+    #   This function is "final" (do not re-implement)
+    #   /sa _getBedTemperature implementation function
+    @pyqtProperty(float, notify = bedTemperatureChanged)
```

Comment:
I actually realised the other day that this is not necessary if we separate the property declaration from the getter function, since then you can safely override the getter function without breaking the property. For example:

`
def getBedTemperature(self):
    raise NotImplementedError()

bedTemperature = pyqtProperty(float, fget = getBedTemperature, notify = bedTemperatureChanged)
`

Unfortunately, since slots are only available as decorators, we cannot use the same trick.